hack/generator/pkg/codegen/pipeline_create_storage_types.go

```diff
@@ -0,0 +1,234 @@
+
+				finalDef := def.WithDescription(descriptionForStorageVariant(d))
+				storageTypes[finalDef.Name()] = finalDef
+			}
+
+			if len(errs) > 0 {
+				err := kerrors.NewAggregate(errs)
+				return nil, err
+			}
+
+			types.AddTypes(storageTypes)
+
+			return types, nil
+		},
+	}
+}
+
+// makeStorageTypesVisitor returns a TypeVisitor to do the creation of dedicated storage types
+func makeStorageTypesVisitor(types astmodel.Types) astmodel.TypeVisitor {
+	factory := &StorageTypeFactory{
+		types: types,
+	}
+
+	result := astmodel.MakeTypeVisitor()
+	result.VisitTypeName = factory.visitTypeName
+	result.VisitObjectType = factory.visitObjectType
+	result.VisitArmType = factory.visitArmType
+
+	factory.visitor = result
+	factory.propertyConversions = []propertyConversion{
+		factory.kubernetesResourceStoragePropertyConversion,
+		factory.defaultStoragePropertyConversion,
+	}
+
+	return result
+}
+
+type StorageTypeFactory struct {
+	types               astmodel.Types
+	propertyConversions []propertyConversion
+	visitor             astmodel.TypeVisitor
+}
+
+// A property conversion accepts a property definition and optionally applies a conversion to make
+// the property suitable for use on a storage type. Conversions return nil if they decline to
+// convert, deferring the conversion to another.
+type propertyConversion = func(property *astmodel.PropertyDefinition, cxt StorageTypesVisitorContext) (*astmodel.PropertyDefinition, error)
+
+func (factory *StorageTypeFactory) visitTypeName(_ *astmodel.TypeVisitor, name astmodel.TypeName, ctx interface{}) (astmodel.Type, error) {
+	vc := ctx.(StorageTypesVisitorContext)
+
+	// Resolve the type name to the actual referenced type
+	actualDefinition, actualDefinitionFound := factory.types[name]
+
+	// Check for property specific handling
+	if vc.property != nil && actualDefinitionFound {
+
+		if et, ok := actualDefinition.Type().(*astmodel.EnumType); ok {
+			// Property type refers to an enum, so we use the base type instead
+			return et.BaseType(), nil
+		}
+	}
+
+	// Map the type name into our storage namespace
+	localRef, ok := name.PackageReference.AsLocalPackage()
+	if !ok {
+		return name, nil
+	}
+
+	storageRef := astmodel.MakeStoragePackageReference(localRef)
+	visitedName := astmodel.MakeTypeName(storageRef, name.Name())
+	return visitedName, nil
+}
+
+func (factory *StorageTypeFactory) visitObjectType(
+	_ *astmodel.TypeVisitor,
+	object *astmodel.ObjectType,
+	ctx interface{}) (astmodel.Type, error) {
+	vc := ctx.(StorageTypesVisitorContext)
+	oc := vc.forObject(object)
+
+	var errs []error
+	properties := object.Properties()
+	for i, prop := range properties {
+		p, err := factory.makeStorageProperty(prop, oc)
+		if err != nil {
+			errs = append(errs, err)
+		} else {
+			properties[i] = p
+		}
+	}
+
+	if len(errs) > 0 {
+		err := kerrors.NewAggregate(errs)
+		return nil, err
+	}
+
+	ot := astmodel.NewObjectType().WithProperties(properties...)
+	return astmodel.NewStorageType(*ot), nil
```

Comment:
Is this why you need the default `VisitStorageType` to accept a returned `StorageType`? I don't think you should though, because you're not going to run this visitor on a storage type. Is that right?